programs/survival/waystones.sc

```diff
@@ -0,0 +1,310 @@
+        );
+    );
+);
+
+_has_required_structure(pos) -> (
+    if(global_settings:'structure_material' && global_settings:'structure_size', 
+        count = 0;
+        scan(pos, 2,2,2, if( _~(global_settings:'structure_material'), count+=1) );
+        count >= global_settings:'structure_size';
+    ,true);
+);
+
+_has_required_item(p) -> (
+    if(!global_settings:'offering', return(true));
+    slot = inventory_find(p, global_settings:'offering');
+    if(slot,
+        item = inventory_get(p, slot);
+        inventory_set(p, slot, item:1 - 1);
+        true;
+    ,
+        _app_message(p, str('The Waystone requires an %s offering', global_settings:'offering'));
+        sound( 'item.trident.return', p~'pos' );
+        false;
+    );
+);
+
+_warp_player(p, pos, dimension, point)->(
+    if(_has_required_item(p),
+        in_dimension(dimension, 
+            run(str('tp %s %d %d %d', p~'command_name', ...(pos + [0.5, 0, 0.5]) ));
+        );
+        schedule( 2, _(p, pos, dimension)->(
+            in_dimension(dimension, 
+                sound( 'item.trident.thunder', pos );
+                particle('totem_of_undying', p~'pos'+[0,1,0], 100, 1, 0.1);
+            );
+        ), p, pos, dimension);
+        _app_message(p, global_waystones:point:'name');
+    );
+);
+
+__on_player_dies(p) -> (
+    if(global_settings:'clear_waypoints_on_death', 
+        uuid = p~'uuid';
+        _read_player_waypoints(uuid);
+        delete(global_waypoints:uuid);
+        _write_player_waypoints(uuid);
+    ); 
+);
+
+_handle_name_tag(item, pos, p, hand) -> (
+    if(item:0=='name_tag' && item:2:'display',
+        global_waystones:pos:'icon':2 = item:2;
+        global_waystones:pos:'name' = _get_name_from_nbt(item:2);
+        slot = if(hand=='mainhand', p~'selected_slot', -1);
+        inventory_set(p, slot, item:1 - 1);
+        sound( 'block.enchantment_table.use', pos );
+        _app_message(p, 'Waypoint named '+global_waystones:pos:'name');
+    true, false);
+);
+
+_mark_player_waypoint(stone_pos, point_pos, uuid) -> (
+    if(!global_waypoints:uuid, global_waypoints:uuid = {}); 
+    global_waypoints:uuid:stone_pos = map(point_pos, floor(_));
+);
+
+// op's waystone screen
+open_waystones_screen() -> (
+    p = player();
+    icons = map(pairs(global_waystones), [_:0, _:1, _:0 + [1,0,0]] );
+    if(length(icons) > 0,  
+        _print_icons_to_screen(_create_warps_screen(p,'fb All Waystones'), icons);
+    ,
+        _app_message(p, 'No waystones to display');
+    );
+);
+
+// open when player clicks on a lodestone
+_open_waypoint_screen(p, pos, uuid) -> (
+    icons = [];
+    bad_keys = [];
+
+    // merge waypoints and waystones data into icons.
+    for(pairs(global_waypoints:uuid),
+        if( global_waystones:(_:0), 
+            icons += ([_:0,global_waystones:(_:0),_:1]);
+        , 
+            bad_keys += _:0;
+        );
+    );
+
+    // Clear player waypoints with missing waystone entires
```

Comment:
```suggestion
    // Clear player waypoints with missing waystone entries
```

Also talking about clearing entries, I've noticed that waypoints destroyed by non-direct player actions (such as explosions) are kept _and_ working. IMO they should be removed, maybe not every time someone opens the menu if it eats performance given it has to load the area, but it could be checked before tp at least.

Reply:
Thats a very good point. I had not even thought about explosions. Ill add a check before the player teleports.

Reply:
I am putting this feature on hold for that fix for the screen callback position bug.